### ISSUE TITLE
Issues with ImageMagick sourcecode

### ISSUE THREAD

**Issue by giulianobelinassi** [SUSE/clang-extract#94]
Found by @5c4lar

```
typedef void (*XtErrorHandler)(
  char *
);
extern void XtSetErrorHandler(
    XtErrorHandler __attribute((noreturn))
);

void f() {
    XtSetErrorHandler(((void*)0));
}
```

```
typedef long unsigned int size_t;
typedef void
  *(*AcquireMemoryHandler)(size_t) ,
  (*DestroyMemoryHandler)(void *),
  *(*ResizeMemoryHandler)(void *,size_t) ,
  *(*AcquireAlignedMemoryHandler)(const size_t,const size_t),
  (*RelinquishAlignedMemoryHandler)(void *);

extern void
  GetMagickMemoryMethods(AcquireMemoryHandler *,ResizeMemoryHandler *, DestroyMemoryHandler *),
  *ResetMagickMemory(void *,int,const size_t);

void f()
{
  (void) ResetMagickMemory((void*)0,0,0);
}
```
and
```
typedef long unsigned int size_t;

/** clang-extract: from source.c:10918:1 */
typedef struct _MagickWand
  MagickWand;

/** clang-extract: from source.c:10935:1 */
typedef enum
{
  MagickFalse = 0,
  MagickTrue = 1
} MagickBooleanType;

/** clang-extract: from source.c:11193:1 */
typedef enum
{
  UndefinedChannel = 0x0000,
  RedChannel = 0x0001,
  GrayChannel = 0x0001,
} ChannelType;

/** clang-extract: from source.c:16204:1 */
extern __attribute__ ((visibility ("default"))) MagickBooleanType
  MagickAdaptiveBlurImage(MagickWand *,const double,const double),
  MagickAdaptiveResizeImage(MagickWand *,const size_t,const size_t),
  MagickAdaptiveSharpenImage(MagickWand *,const double,const double);

/** clang-extract: from source.c:17643:1 */
MagickBooleanType f(MagickWand *wand, const ChannelType channel, const double radius, const double sigma) {
 MagickBooleanType status;
 status = MagickAdaptiveSharpenImage(wand, radius, sigma);
 return status;
}
```

**Comment by giulianobelinassi** [comment]
This case:
```
typedef void (*XtErrorHandler)(
  char *
);
extern void XtSetErrorHandler(
    XtErrorHandler __attribute((noreturn))
);

void f() {
    XtSetErrorHandler(((void*)0));
}
```
Is a bug in clang's AST. It doesn't reference `XtErrorHandler` on `XtSetErrorHandler`'s parameter:
```
FunctionProtoType 0x5210000720e0 'void (void (*)(char *) __attribute__((noreturn)))' cdecl
|-BuiltinType 0x521000014170 'void'
`-PointerType 0x521000071fd0 'void (*)(char *) __attribute__((noreturn))'
  `-ParenType 0x521000071f70 'void (char *) __attribute__((noreturn))' sugar
    `-FunctionProtoType 0x521000071f30 'void (char *) __attribute__((noreturn))' noreturn cdecl
      |-BuiltinType 0x521000014170 'void'
      `-PointerType 0x521000014d10 'char *'
        `-BuiltinType 0x5210000141b0 'char'
```
Dropping the `__attribute((noreturn))` results in:
```
FunctionProtoType 0x521000071ff0 'void (XtErrorHandler)' cdecl
|-BuiltinType 0x521000014170 'void'
`-ElaboratedType 0x521000071f00 'XtErrorHandler' sugar
  `-TypedefType 0x521000071ed0 'XtErrorHandler' sugar
    |-Typedef 0x521000071e78 'XtErrorHandler'
    `-PointerType 0x521000071e10 'void (*)(char *)'
      `-ParenType 0x521000071db0 'void (char *)' sugar
        `-FunctionProtoType 0x521000071d70 'void (char *)' cdecl
          |-BuiltinType 0x521000014170 'void'
          `-PointerType 0x521000014d10 'char *'
            `-BuiltinType 0x5210000141b0 'char'
```

**Comment by giulianobelinassi** [comment]
This case:
```
typedef long unsigned int size_t;
typedef void
  *(*AcquireMemoryHandler)(size_t) ,
  (*DestroyMemoryHandler)(void *),
  *(*ResizeMemoryHandler)(void *,size_t) ,
  *(*AcquireAlignedMemoryHandler)(const size_t,const size_t),
  (*RelinquishAlignedMemoryHandler)(void *);

extern void
  GetMagickMemoryMethods(AcquireMemoryHandler *,ResizeMemoryHandler *, DestroyMemoryHandler *),
  *ResetMagickMemory(void *,int,const size_t);

void f()
{
  (void) ResetMagickMemory((void*)0,0,0);
}
```

Is the same issue as in #95. We certainly will need to check and analyze decls that have the same BeginLoc. One can workaround this by calling clang-extract with `-DCE_EXTRACT_FUNCTIONS=GetMagickMemoryMethods,ResetMagickMemory`

**Comment by giulianobelinassi** [comment]
The same thing happens here:
```
typedef long unsigned int size_t;

/** clang-extract: from source.c:10918:1 */
typedef struct _MagickWand
  MagickWand;

/** clang-extract: from source.c:10935:1 */
typedef enum
{
  MagickFalse = 0,
  MagickTrue = 1
} MagickBooleanType;

/** clang-extract: from source.c:11193:1 */
typedef enum
{
  UndefinedChannel = 0x0000,
  RedChannel = 0x0001,
  GrayChannel = 0x0001,
} ChannelType;

/** clang-extract: from source.c:16204:1 */
extern __attribute__ ((visibility ("default"))) MagickBooleanType
  MagickAdaptiveBlurImage(MagickWand *,const double,const double),
  MagickAdaptiveResizeImage(MagickWand *,const size_t,const size_t),
  MagickAdaptiveSharpenImage(MagickWand *,const double,const double);

/** clang-extract: from source.c:17643:1 */
MagickBooleanType f(MagickWand *wand, const ChannelType channel, const double radius, const double sigma) {
 MagickBooleanType status;
 status = MagickAdaptiveSharpenImage(wand, radius, sigma);
 return status;
}
```
Here, `size_t` is not added to the closure because `MagickAdaptiveSharpenImage` does not reference it.

**Comment by giulianobelinassi** [comment]
The first case is fixed by #97.
